### PR TITLE
Classifier drawing tools: simplify SVGContext

### PR DESCRIPTION
**Pull request by eatyourgreens**
Remove the imageViewer ref from SingleImageViewerContainer and MultiFrameViewerContainer. Instead, reference the svg DOM node from SingleImageViewer. Pass that node down to individual drawing components in SVGContext.

~~Remove `getScreenCTM` from the context since any SVG element should have a `getScreenCTM` method.~~ Messing with the transformation matrix breaks tools that use `rotate` transforms, so I've left the context alone.

Package:
lib-classifier

# Review Checklist

## General

- [ ] Are the tests passing locally and on Travis?
- [ ] Is the documentation up to date?

## Components
- [ ] Has a storybook story been created or updated?
- [ ] Is the component accessible? 
  - [ ] Can it be used with a screen reader? [BBC guide to testing with VoiceOver](https://bbc.github.io/accessibility-news-and-you/accessibility-and-testing-with-voiceover-os.html)
  - [ ] Can it be used from the keyboard? [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - [ ] Is it passing accessibility checks in the storybook?

## Apps

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `yarn panic && yarn bootstrap` or `docker-compose up --build` and app works as expected?

## Publishing

- [ ] Is the changelog updated?
- [ ] Are the dependencies updated for apps and libraries that are using the newly published library?

## Post-merging

- [ ] Did the app deploy to https://frontend.preview.zooniverse.org/projects/:project-name/:owner or https://frontend.preview.zooniverse.org/about?
- [ ] Is the new feature working or bug now fixed?
  - [ ] Is there a Talk or blog post written to announce the new feature(s)?
- [ ] Is the design working across browsers (Firefox, Chrome, Edge, Safari) and mobile?
  - [ ] Is this approved by our designer?
- [ ] Is this ready for production deployment?

### CODE REVIEW

**Comment by eatyourgreens** [comment]
Tests are broken but local testing looks good, trying some demo classifications from a test project. Subject dimensions are still being captured.

**Review by shaunanoordin — COMMENTED** [review]
## PR Review (Ongoing)

Context: Most of our Subject Viewers require references to the `<svg>` element in the HTML (and/or related sub-elements in the SVG), usually to calculate the dimensions of the actual subject image and/or the position of mouse events on said image, relative to its actual visual presentation (zoomed in/out, panned left/right/etc) on the web browser.

Goals: this PR attempts to simplify how those references are passed around, since we've been encountering issues where the complicated nest of conditional renderings + refs + forwarded refs could, under specific conditions, screw up these references. (see #2221, #2241, #2251 for additional reading)

Notes:
- ⚠️  One major thing to note is that the [SVG Point](https://developer.mozilla.org/en-US/docs/Web/API/SVGPoint) (e.g. in draggable.js's `scg.createSVGPoint) that's essential to a lot of our component code is **deprecated**
  - This PR will not address this issue (since SVGPoint is so essential for everything to work, and this PR is supposed to be a refactor only) but it's worth noting as a semi-urgent Future TODO.

### Status

Just finished code read (looks good) and I'm now in the process of testing.

TODO:
- Test that drawing tools work on a Single Image Viewer workflow (I Fancy Cats)
- Test that drawing tools work on a Multi Frame Viewer workflow (Davy Notebooks)
- Test that drawing tools work on a Single Video Viewer (Solar Jets?)
- Test that the Subject Group Viewer works. (SURVOS survey)
- Confirm that the fix for issue #2241 doesn't regress. (Shouldn't be an issue.)
- Check that Classifications are submitted with the correct image dimensions. (The "viewer size" metadata)
- Check that the Annotations of Classifications have the correct coordinates! (Shaun, use that test project with the grid Subject)
  - Double-check with panned/zoomed images!
  - Double-check with wonky browser screen sizes, if you have the time
- Check SAFARI.

**Comment by shaunanoordin** [comment]
## PR Review (Update)

I need to make a note of this because I keep alt-tabbing into too many goshdang windows

### Testing

Testing will be done on macOS 10.12 + Chrome 91 (with a Safari exception later), on localhost.

**Scenario 1:** can I draw on the Single Image Viewer?
URL: [I Fancy Cats](http://localhost:3000/projects/brooke/i-fancy-cats/classify/workflow/693?env=staging)
Testing:
- [ ] Classifier should load normally.
- [ ] Pan and zoom should work.
- [ ] Proceed to the drawing step. Draw some Points.
- [ ] Submit Classification. Check Annotations and metadata: Subject Dimensions.

**Scenario 2:** can I transcribe text on the Multi-Frame Viewer?
URL: [Davy Notebooks](http://localhost:3000/projects/humphrydavy/davy-notebooks-project?env=production) (not a direct link to the WF, since D-notes's WF keeps changing)
Testing:
- [ ] Classifier should load normally.
- [ ] Pan and zoom should work.
- [ ] Proceed to the drawing step. Transcribe some lines of text.
- [ ] Submit Classification. Check Annotations and metadata: Subject Dimensions.

**Scenario 3:** can I draw on the Single Video Viewer?
URL: [Erik's Video Testing Project](http://localhost:3000/projects/eostlund/video-project/classify/workflow/16843?env=production)
Testing:
- [ ] Classifier should load normally.
- [ ] Proceed to the drawing step. Draw some Points.
- [ ] Submit Classification. Check Annotations and metadata: Subject Dimensions.

**Scenario 4** can I work on the Subject Group Viewer (aka grid tool)?
URL: [SURVOS Testing](http://localhost:3000/projects/darkeshard/survos-testing-2020-subject-group-viewer/classify/workflow/3412?env=staging)
- [ ] Classifier should load normally.
- [ ] Pan and zoom should work.
- [ ] Select and/or unselect some cells in the grid.
- [ ] Submit Classification. Check Annotations and metadata: Subject Dimensions.

**Scenario 5** is issue #2241 still a thing?
URL: [Davy Notebooks](http://localhost:3000/projects/humphrydavy/davy-notebooks-project?env=production) (not a direct link to the WF, since D-notes's WF keeps changing)
- Open classifier page. Transcribe some text. Navigate to project home page. Navigate back to classifier page.
- [ ] You should be able to modify existing transcription lines, and add new ones.

**Scenario 6** are you sure that all the annotations have correct coordinates?
URL: [Shaun's Precise 1000x1000px Grid For Testing Coordinates](http://localhost:3000/projects/darkeshard/prototype-2015/classify/workflow/18372?env=production)
- Draw some Rectangles or Points, paying close attention to the intended coordinates and dimensions. 
- [ ] Submit Classification. Check Annotations' coordinates and dimensions.
- Resize the window. Pan and zoom and rotate.
- Draw some Rectangles or Points, paying close attention to the intended coordinates and dimensions. 
- Resize the window. Pan and zoom and rotate AGAIN.
- [ ] Submit Classification. Check Annotations' coordinates and dimensions.

**Scenario 7** Safari?
URL: [Davy Notebooks](http://localhost:3000/projects/humphrydavy/davy-notebooks-project?env=production) (not a direct link to the safari, since safari's safari keeps safari)
- [ ] Safari 12? (my machine)
- [ ] Safari 14? (browserstack)
- Safari safari.

### Status

Testing ongoing. If all the checkboxes above are ticked, everything is hunky dory.

**Review by shaunanoordin — CHANGES_REQUESTED** [review]
## PR Review (Update)

I've just finished testing (see https://github.com/zooniverse/front-end-monorepo/pull/2255#issuecomment-866327326) and here are my findings:

- _Single Image Viewer, Multi-Frame Viewer, and Subject Group Viewer_ all working fine. 👍 
- _Most_ Drawing Tool annotations/Drawing Marks, _especially Transcription Text Lines,_ are working correctly. 👍 
- ❗ Everything looks good EXCEPT for a new issue introduced to Rotate Rectangle and Ellipsis.
- ❓ _Single Video Viewer_ couldn't be fully tested. Might be worth tagging @ErikOstlund in on this as an FYI. (Erik, I might need to talk to you about spinning off another WF with a video viewer on my project, so I can test things without interrupting your TemporalRotateRectangle workflow.)

**New Issue:** "drag handle to rotate" on certain Drawing Marks are now broken.
- RotateRectangle and Ellipse drawing marks have a "draggable handle" that one can drag to rotate the mark.
- With this PR, attempting such a drag-to-rotate action unfortunately causes the rotation to go haywire.
- Conversely, _drag-to-move_ (tested on Point, Rectangle, RotateRectangle, and Ellipsis) works fine. _Drag-to-resize_ (tested on Rectangle, RotateRectangle) works fine. _Drag-to-create_ (Rectangle, RotateRectangle, Ellipsis) works fine. It's only rotation!

Replication:
- You need a project with a Single Image Viewer, with a Drawing Task, with the RotateRectangle and/or Ellipsis tools.
- Baseline URL: https://frontend.preview.zooniverse.org/projects/darkeshard/prototype-2015/classify/workflow/18372?env=production
- Testing URL: http://localhost:3000/projects/darkeshard/prototype-2015/classify/workflow/18372?env=production
- Draw a RotateRectangle.
- Drag the handle to rotate
  - Expectation (confirmed on baseline): the drag handle should follow the mouse, causing the rectangle to rotate to "face" the mouse cursor.
  - Actual: the rotation goes all over the page, with no regard to the actual or relative position of the mouse cursor.

Analysis? Theories? (more like random thoughts, actually. Read the following with a grain of salt.):
- I think the issue is with changes to draggable.js?
- It's possible that the _coordinate offsets_ being passed to the "rotate me!" function isn't using the same _coordinate-references_ as the component's. (e.g. myRotateRectangle has x,y coords _relative to the parent <svg>,_ but the the "drag to rotate" event is passing x,y coords _relative to myRotateRectangle._ ) 
- Maybe the drag-to-move and drag-to-resize and drag-to-create aren't affected because those ops just require delta-x and delta-y (i.e. the difference from the start of drag to the current drag position) to perform flat horizontal and vertical transforms, so the difference in coordinate-references just cancels out. An Ellipse's "drag-to-rotate" function however, DOES require those coordinate-references to be the same to calculate angle.

### Status

We just need to figure out this new issue; everything else looks good.

Jim, I'll dig deeper into the Ellipse and RotateRectangle mechanics tomorrow afternoon, unless you figure it out first. 👍

**Comment by eatyourgreens** [comment]
Weird. The changes to `getScreenCTM` in the second commit might not correct for rotation transforms.

**Comment by eatyourgreens** [comment]
I’d be absolutely fine with removing that second commit if it’s causing problems.

**Comment by eatyourgreens** [comment]
@shaunanoordin I've removed that second commit. The rotated drawing tools break unless you use `getScreenCTM` from the top-level transform.

Dragging a rotated image seems to be broken on master so 🤷‍♂️ .

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js [hunk 1]
```diff
@@ -64,7 +65,8 @@ function InteractionLayer({
     const svgPoint = svg.createSVGPoint()
     svgPoint.x = event.clientX
     svgPoint.y = event.clientY
-    const svgEventOffset = svgPoint.matrixTransform(getScreenCTM().inverse())
+    const inverseTransform = canvas.current?.getScreenCTM().inverse()
+    const svgEventOffset = svgPoint.matrixTransform(inverseTransform)
```
[the remark this answers is not in the document]
Comment (eatyourgreens) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js]: Something in this commit broke `rotate` transforms, so I've removed it.

**Review by shaunanoordin — APPROVED** [review]
## PR Review

LGTM! Thanks for reverting that commit Jim, that seems to have fixed the issue.

- I've tested Scenarios 1-2, and 4-7 again, and everything looks good 👍 
- RotateRectangle and Ellipsis now rotate gracefully, like an autumn leaf spinning in the wind, or a Dutch windmill, or a cow flying in a tornado. Wait, I might have lost the plot. Basically, _it works._

Giving this PR a big ol' thumbs up! 👍

**Comment by shaunanoordin** [comment]
I just realised what's the better way of describing this issue:

> shaunanoordin I've removed that second commit. The rotated drawing tools break **unless you use getScreenCTM from the top-level transform.**

Basically, right now the Classifier is passing around `<svg>` and `<g id="transformation_layer">.getScreenCTM()`  via SVGContext. (See: SingleImageViewer.js)

That might be a bit confusing because it's passing the _reference to one element,_ and the _function to a DIFFERENT element._ 

It might make a bit more sense for the SVGContext to instead pass around `<svg>` and `<g id="transformation_layer">` instead (i.e. two references to different elements) for consistency; this would make it super clear to the context-consuming components (i.e. draggable.js, InteractionLayer.js) _which element_ it's pulling getScreenCTM() from.

**Pull request merged**